src/cpp/core/include/core/http/AsyncServerImpl.hpp

```diff
@@ -325,11 +332,30 @@ class AsyncServerImpl : public AsyncServer, boost::noncopyable
 
          // call the appropriate handler to generate a response
          std::string uri = pRequest->uri();
-         AsyncUriHandlerFunction handler = uriHandlers_.handlerFor(uri);
-         if (handler)
+         AsyncUriHandler handler = uriHandlers_.handlerFor(uri);
+         AsyncUriHandlerFunction handlerFunc = handler.function();
+
+         if (handlerFunc)
          {
+            if (!handler.isProxyHandler())
+            {
+               // check to ensure the request is for a supported method
+               // proxy handlers do not perform this checking as we
+               // flow all traffic like a proxy
+               const std::string& method = pRequest->method();
+               if (method != "GET" &&
+                   method != "PUT" &&
+                   method != "POST")
+               {
+                  // invalid method - fail out
+                  LOG_ERROR_MESSAGE("Invalid method " + method + " requested for uri: " + pRequest->uri());
+                  pConnection->response().setStatusCode(http::status::MethodNotAllowed);
+                  return;
+               }
+            }
+
             // call the handler
-            handler(pAsyncConnection) ;
+            handlerFunc(pAsyncConnection) ;
          }
          else if (defaultHandler_)
```

Comment:
Do we also need to guard requests made to the default handler, and when no handler is registered? (e.g. for GET a 404 seems appropriate, but perhaps not for other verbs)